keps/sig-api-machinery/4006-transition-spdy-to-websockets/README.md

```diff
@@ -426,6 +449,93 @@ translation proxy terminates the WebSocket connection, and it de-multiplexes the
 various streams in order to pass the data on to a SPDY connection, which continues
 upstream (to Kubelet and eventually the container runtime).
 
+### Background: `PortForward` Subprotocol
+
+The `PortForward` subprotocol is used to implement `kubectl port-forward`, and it differs
+from the `RemoteCommand` subprotocol in how the multiple streams within the single upgraded
+connection are created. The `RemoteCommand` subprotocol statically creates streams (e.g
+STDIN, STDOUT, etc.) when the connection is created. But the `PortForward` subprotocol
+dynamically creates two streams (a bi-directional data stream and a unidirectional error
```

Comment:
Theoretically it creates multiple bi-directional data streams, since it's possible to forward multiple ports in a single invocation. Each port pair (src:dst) is a single data stream.

Reply:
Is the error stream shared for all pairs or is there an error stream per pair?

Reply:
There is an error stream (and data stream) per **request**. There are two steps for `port-forward`:

1. The `kubectl port-forward <TARGET> [<LOCAL_PORT:REMOTE_PORT>+]`. This step sets up the local ports to listen on and creates the upgraded connection. There are zero streams at this point.

2. A port-forward **request** (e.g. `curl http://localhost:LOCAL_PORT/index.html` in a separate terminal). A data stream and error stream pair are dynamically created over the connection to handle this **request**. Many requests can occur concurrently .

Reply:
I've updated nomenclature from **request** to be **subrequest** where necessary to clarify these subtle differences. And I've added these steps which articulate the difference between these two terms in two different places. In this location (`Background: PortForward Subprotocol`), we are describing the current SPDY implementation as background information. I also describe these steps in the `Proposal: API Server PortForward StreamTranslatorProxy` section with important modifications relevant to the changes being proposed. Please let me know what you think.

Reply:
lgtm